src/models/Visitors.php

```diff
@@ -0,0 +1,344 @@
+
+        if ($referer !== null) {
+            return $this->getPageId($referer);
+        } else {
+            return null;
+        }
+    }
+
+    public function addLastVisitor($visitorId)
+    {
+        if ($this->_statusAddLastVisitor && $this->get($this::SESSION_NAME_KEY, false)) {
+            VisitorSession::updateAll([
+                'last_action_at' => date('Y-m-d H:i:s'),
+                'last_visited_page_id' => $this->getPageId(Yii::$app->request->url)
+            ], ['=', 'id', $visitorId]);
+        }
+    }
+
+    public function addLastSession($sessionId)
+    {
+        if ($this->_statusAddLastSession && $this->get($this::VISITOR_NAME_KEY, false)) {
+            // incorrect
+            Visitor::updateAll([
+                'last_activity_at' => date('Y-m-d H:i:s'),
+                'last_activity_visited_page_id' => $this->getPageId(Yii::$app->request->url)
+            ], ['=', 'id', $sessionId]);
+        }
+    }
+
+
+    /**
+     * @return int
+     */
+    private function sessionLifeTime()
+    {
+        return time() + 30;
+    }
+
+    public function run()
+    {
+
+        list($visitorId, $sessionId) = $this->init();
+
+        if ($visitorId !== null && $sessionId !== null) {
+
+//             if visitor first logged. add user id
+            if (
+                $this->_statusAddUser
+                && !$this->get($this::USER_IF_AUTO, false)
+                && Yii::$app->user->getId() !== null
+            ) {
+                Visitor::updateAll([
+                    'user_id' => Yii::$app->user->getId()
+                ], ['=', 'id', $visitorId]);
+
+                $this->set($this::USER_IF_AUTO, true);
+            }
+
+
+            $this->addLastVisitor($visitorId);
+
+            $this->addLastSession($sessionId);
+        }
+
+    }
+
+
+    /**
+     * @return string
+     */
+    public function getIpUser()
+    {
+        return Yii::$app->getRequest()->getUserIP();
+    }
+
+    /**
+     * @param $link
+     */
+    public function getPageId($link)
+    {
+        if ($pageId = Yii::$app->cache->get($link)) {
+            return $pageId;
+        } else {
+            $pageId = $this->getPageIdFromDB($link);
+            Yii::$app->cache->set($link, $pageId, 24 * 60 * 60);
+            return $pageId;
+        }
+
+    }
+
+    /**
+     * @return bool|string
+     */
+    public function startedAt()
+    {
+        return date('Y-m-d H:i:s');
+    }
+
+    /**
+     * @return bool|string
+     */
+    public function lastActionAt()
+    {
+        return date('Y-m-d H:i:s');
+    }
+
+    /**
+     * @return bool|string
+     */
+    public function firstActivityAt()
+    {
+        return date('Y-m-d H:i:s');
+    }
+
+    /**
+     * @return bool|string
+     */
+    public function lastActivityAt()
+    {
+        return date('Y-m-d H:i:s');
+    }
+
+    /**
+     * @return int
+     */
+    protected function timeExpire()
+    {
+        return time() + 24 * 60 * 60;
```

Comment:
why not to use some configurable public variable in session and plus it to the time?